pilot/pkg/proxy/envoy/v2/eds.go

```diff
@@ -142,9 +195,175 @@ func newEndpoint(e *model.NetworkEndpoint) (*endpoint.LbEndpoint, error) {
+
+	portMap, f := push.ServicePort2Name[string(hostname)]
+	if !f {
+		return s.updateCluster(push, clusterName, edsCluster)
+	}
+	portName, f := portMap[uint32(port)]
+	if !f {
+		return s.updateCluster(push, clusterName, edsCluster)
+	}
+
+	// The service was never updated - do the full update
+	se, f := s.EndpointShardsByService[string(hostname)]
+	if !f {
+		return s.updateCluster(push, clusterName, edsCluster)
+	}
+
+	cnt := 0
+	localityEpMap := make(map[string]*endpoint.LocalityLbEndpoints)
+
+	// The shards are updated independently, now need to filter and merge
+	// for this cluster
+	for _, es := range se.Shards {
+		for _, el := range es.Entries {
+			if portName != el.ServicePortName {
+				continue
+			}
+			// Port labels
+			if !labels.HasSubsetOf(model.Labels(el.Labels)) {
+				continue
+			}
+			cnt++
+
+			// TODO: Need to accommodate region, zone and subzone. Older Pilot datamodel only has zone = availability zone.
+			// Once we do that, the key must be a | separated tupple.
+			locality := (el.Labels)[model.AZLabel] // may be ""
+			locLbEps, found := localityEpMap[locality]
+			if !found {
+				locLbEps = &endpoint.LocalityLbEndpoints{
+					Locality: &core.Locality{
+						Zone: locality,
+					},
+				}
+				localityEpMap[locality] = locLbEps
+			}
+			if el.EnvoyEndpoint == nil {
+				el.EnvoyEndpoint = serviceEntry2Endpoint(el.UID, el.Family, el.Address, el.EndpointPort)
+			}
+			locLbEps.LbEndpoints = append(locLbEps.LbEndpoints, *el.EnvoyEndpoint)
+		}
+	}
+	locEps := make([]endpoint.LocalityLbEndpoints, 0, len(localityEpMap))
+	for _, locLbEps := range localityEpMap {
+		locEps = append(locEps, *locLbEps)
+	}
+
+	if cnt == 0 {
+		push.Add(model.ProxyStatusClusterNoInstances, clusterName, nil, "")
+		//adsLog.Infof("EDS: no instances %s (host=%s ports=%v labels=%v)", clusterName, hostname, p, labels)
+	}
+	edsInstances.With(prometheus.Labels{"cluster": clusterName}).Set(float64(cnt))
+
+	// There is a chance multiple goroutines will update the cluster at the same time.
+	// This could be prevented by a lock - but because the update may be slow, it may be
+	// better to accept the extra computations.
+	// We still lock the access to the LoadAssignments.
+	edsCluster.mutex.Lock()
+	defer edsCluster.mutex.Unlock()
+
+	edsCluster.LoadAssignment = &xdsapi.ClusterLoadAssignment{
+		ClusterName: clusterName,
+		Endpoints:   locEps,
+	}
+	if len(locEps) > 0 && edsCluster.NonEmptyTime.IsZero() {
+		edsCluster.NonEmptyTime = time.Now()
+	}
+	return nil
+}
+
+// updateServiceShards will list the endpoints and create the shards.
+// This is used to reconcile and to support non-k8s registries (until they migrate).
+// Note that aggreaged list is expensive (for large numbers) - we want to replace
+// it with a model where DiscoveryServer keeps track of all endpoint registries
+// directly, and calls them one by one.
+func (s *DiscoveryServer) updateServiceShards(push *model.PushContext) error {
```

Comment:
this will read much better if its UpdateServiceInstanceIndices.. ServiceShards seem to indicate that you are partitioning services into one or more groups

Reply:
And that's exactly what I am doing - partition services into small groups ( shards ) that can be updated as a unit.

Reply:
In other words:  updating a deployment should not result in appending to a big list and concatenations and allocations.  Just update a map.

Reply:
What is the criteria for grouping services in to Shards? how do you infer the relationship between services? And why group services in to shards, when a change in the service will effectively trigger changes to bunch of listeners, clusters, routes, etc. across the board? Also, the comment elsehwere says ServiceShard is bunch of endpoint groups

```
// ServiceShards holds the set of endpoint shards of a service. Registries update
// individual shards incrementally. The shards are aggregated and split into
// clusters when a push for the specific cluster is needed.
type ServiceShards struct {
 	// Shards is used to track the shards. EDS updates are grouped by shard.
	// Current implementation uses the registry name as key - in multicluster this is the
	// name of the k8s cluster, derived from the config (secret).
	Shards map[string]*EndpointShard
 	// ServiceAccounts has the concatenation of all service accounts seen so far in endpoints.
	// This is updated on push, based on shards. If the previous list is different than
	// current list, a full push will be forced, to trigger a secure naming update.
	// Due to the larger time, it is still possible that connection errors will occur while
	// CDS is updated.
	ServiceAccounts map[string]bool
}
```

While the comment above ```And that's exactly what I am doing - partition services into small groups ( shards ) that can be updated as a unit.``` implies something totally different. 
IF its a group of services, then the ServiceShards struct above implies that multiple such services share same endpoint and same service account, which does not really seem right. 

And now I am confused ..

Reply:
Ok, I see what confused you. It's 'shards of a service'... 

Will rename them - a big pain since the merge to master is manual, just few extra days of pain...